rfc/004-reverse-sync.md

```diff
@@ -0,0 +1,218 @@
+and [State sync](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-042-state-sync.md)
+meant nodes no longer needed a complete history of the blockchain. This
+introduced some challenges of its own which were covered and subsequently
+tackled with [RFC-001](https://github.com/tendermint/spec/blob/master/rfc/001-block-retention.md).
+The RFC allowed applications to set a block retention height; an upper bound on
+what blocks would be pruned. However nodes who state sync past this upper bound
+(which is necessary as snapshots must be saved within the trusting period for
+the assisting light client to verify) have no means of backfilling the blocks
+to meet the retention limit. This could be a problem as nodes who state sync and
+then eventually switch to consensus may not have the block and validator
+history to verify evidence causing them to panic if they see 2/3 commit on what
+the node believes to be an invalid block.
+
+Furthermore, having the capability to backfill prior blocks also allows nodes with
+truncated history to become archive nodes which may be beneficial for the
+network.                                            
+
+## Proposal
+
+A backfill mechanism can simply be defined as an algorithm for fetching,
+verifying and storing, blocks and other state data (`ValidatorSet`'s
+`ConsensusParams`'s and `ABCIResponses`'s) of a height prior to the current
+base of the node's blockchain. In matching the terminology used for other
+data retrieving protocols (i.e. fast sync and state sync), we 
+call this method **ReverseSync**. Verification and design will be 
+addressed later in the RFC. First we cover where the mechanism would be deployed.
+
+There are two places where a backfill mechanism would be required:
+1. Upon startup via state sync where nodes must meet the prerequisite history
+before participating in consensus
+2. Where application specifies that the `retain_height` needs to be lower than
+the current blockchain base.
+
+### ReverseSync after State Sync
+
+A node, seeking to state sync, will find an adequate snapshot and offer it to
+the application with the following format:
+
+```proto
+message RequestOfferSnapshot {
+  Snapshot snapshot = 1;  // snapshot offered by peers
+  bytes    app_hash = 2;  // light client-verified app hash for snapshot height
+}
+```
+
+where `Snapshot` contains the new field `backfill_height` :
+
+```proto
+message Snapshot {
+  uint64 height          = 1;  // The height at which the snapshot was taken
+  uint32 format          = 2;  // The application-specific snapshot format
+  uint32 chunks          = 3;  // Number of chunks in the snapshot
+  bytes  hash            = 4;  // Arbitrary snapshot hash, equal only if identical
+  bytes  metadata        = 5;  // Arbitrary application metadata
+  uint64 backfill_height = 6;  // Height to backfill blocks from before starting application (inclusive)
```

Comment:
That's a valid concern. In RFC001 on pruning, it was decided to create a retainHeight variable and leave it up to the application to decide how much they wanted to prune as it was believed that they would have access to the most information (there may be some app specific invariants that might mean for example that they would need to keep the last 2 months of blocks). Basically we already decided to put the responsibility in the hand of the app. This is just a continuation of this.

Reply:
Something I haven't considered is whether this could be used as a sort of attack. Like what if I create a snapshot that insists other nodes must reverse sync back to genesis. Perhaps we're also implying that this can be verified. 

Another idea may be to deliver all the chunks to the application first and then after the last chunk, ask the application what the retain height should be and then reverse sync to there